objects/src/accounts/delta/vault.rs

```diff
@@ -15,314 +18,466 @@ use crate::{
+        let delta = Self(map);
         delta.validate()?;
 
         Ok(delta)
     }
 
-    /// Checks whether this vault delta is valid.
+    /// Adds a new fungible asset to the delta.
+    ///
+    /// # Errors
+    /// Returns an error if the delta would overflow.
+    pub fn add(&mut self, asset: FungibleAsset) -> Result<(), AccountDeltaError> {
+        let amount: i64 = asset.amount().try_into().expect("Amount it too high");
+        self.add_delta(asset.faucet_id(), amount)
+    }
+
+    /// Removes a fungible asset from the delta.
     ///
     /// # Errors
-    /// Returns an error if:
-    /// - The number added assets is greater than [u16::MAX].
-    /// - The number of removed assets is greater than [u16::MAX].
-    /// - The same asset was added more than once, removed more than once, or both added and
-    ///   removed.
-    pub fn validate(&self) -> Result<(), AccountDeltaError> {
-        if self.added_assets.len() > u16::MAX as usize {
-            return Err(AccountDeltaError::TooManyAddedAsset {
-                actual: self.added_assets.len(),
-                max: u16::MAX as usize,
-            });
-        } else if self.removed_assets.len() > u16::MAX as usize {
-            return Err(AccountDeltaError::TooManyRemovedAssets {
-                actual: self.removed_assets.len(),
-                max: u16::MAX as usize,
-            });
+    /// Returns an error if the delta would overflow.
+    pub fn remove(&mut self, asset: FungibleAsset) -> Result<(), AccountDeltaError> {
+        let amount: i64 = asset.amount().try_into().expect("Amount it too high");
+        self.add_delta(asset.faucet_id(), -amount)
+    }
+
+    /// Returns true if this vault delta contains no updates.
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    /// Returns an iterator over the (key, value) pairs of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&AccountId, &i64)> {
+        self.0.iter()
+    }
+
+    /// Merges another delta into this one, overwriting any existing values.
+    ///
+    /// The result is validated as part of the merge.
+    ///
+    /// # Errors
+    /// Returns an error if the result did not pass validation.
+    pub fn merge(&mut self, other: Self) -> Result<(), AccountDeltaError> {
+        // Merge fungible assets.
+        //
+        // Track fungible asset amounts - positive and negative. `i64` is not lossy while
+        // fungibles are restricted to 2^63-1. Overflow is still possible but we check for that.
+
+        for (&faucet_id, &amount) in other.0.iter() {
+            self.add_delta(faucet_id, amount)?;
         }
 
-        // make sure all added assets are unique
-        for (pos, asset) in self.added_assets.iter().enumerate() {
-            if self.added_assets[..pos].iter().any(|a| a.is_same(asset)) {
-                return Err(AccountDeltaError::DuplicateVaultUpdate(*asset));
-            }
+        self.validate()
```

Comment:
Why do we need this validation here? The only thing `.validate()` does is make sure that account ID are fungible faucet IDs. But here we already know that for `self` and for `other` all account IDs are fungible faucet IDs.